File: src/dependencies/crazyflie_hardware/src/crtp_cpp/src/crtp_cpp/packer/basic_commander_packer.cpp
#include "crtp_cpp/packer/basic_commander_packer.hpp"
#include <cstring>

#define PORT_BASIC_COMMANDER 0x03
#define CHANNEL 0x0

BasicCommanderPacker::BasicCommanderPacker()
    : CrtpPacker(PORT_BASIC_COMMANDER) {}

CrtpPacket BasicCommanderPacker::prepare_packet(const std::vector<uint8_t>& data) {
    return CrtpPacker::prepare_packet(CHANNEL, data);
}

CrtpPacket BasicCommanderPacker::send_setpoint(float roll, float pitch, float yawrate, uint16_t thrust) {
    std::vector<uint8_t> data(4 + 4 + 4 + 2);

    std::memcpy(&data[0], &roll, 4);
    std::memcpy(&data[4], &pitch, 4);
    std::memcpy(&data[8], &yawrate, 4);
    std::memcpy(&data[12], &thrust, 2);

    return prepare_packet(data);
}

File: src/dependencies/crazyflie_hardware/src/crtp/crtp/packers/basic_commander_packer.py
from .packer import Packer
from .crtp_packer import CrtpPacker

import struct

from typing import Callable


class BasicCommanderPacker(Packer):
    PORT_BASIC_COMMANDER = 0x03

    CHANNEL = 0x0

    def __init__(self, crtp_packer_factory: Callable[[int], CrtpPacker]):
        super().__init__(crtp_packer_factory, self.PORT_BASIC_COMMANDER)

    def send_setpoint(self, roll, pitch, yawrate, thrust):
        data = struct.pack("<fffH", roll, pitch, yawrate, thrust)
        return self._prepare_packet(data=data)
